crates/red_knot_python_semantic/resources/mdtest/protocols.md

```diff
@@ -1,9 +1,1015 @@
+
+@dataclass
+class MutableDataclass:
+    x: int
+
+static_assert(is_subtype_of(MutableDataclass, HasX))  # error: [static-assert-error]
+static_assert(is_assignable_to(MutableDataclass, HasX))  # error: [static-assert-error]
+
+@dataclass(frozen=True)
+class ImmutableDataclass:
+    x: int
+
+static_assert(not is_subtype_of(ImmutableDataclass, HasX))
+static_assert(not is_assignable_to(ImmutableDataclass, HasX))
+
+class NamedTupleWithX(NamedTuple):
+    x: int
+
+static_assert(not is_subtype_of(NamedTupleWithX, HasX))
+static_assert(not is_assignable_to(NamedTupleWithX, HasX))
+```
+
+However, a type with a read-write property `x` *does* satisfy the `HasX` protocol. The `HasX`
+protocol only specifies what the type of `x` should be when accessed from instances; instances of
+`XProperty` in the below example have a mutable attribute `x` of type `int`:
+
+```py
+class XProperty:
+    _x: int
+
+    @property
+    def x(self) -> int:
+        return self._x
+
+    @x.setter
+    def x(self, x: int) -> None:
+        self._x = x**2
+
+static_assert(is_subtype_of(XProperty, HasX))  # error: [static-assert-error]
+static_assert(is_assignable_to(XProperty, HasX))  # error: [static-assert-error]
+```
+
+Attribute members on protocol classes are allowed to have default values, just like instance
+attributes on other classes. Similar to nominal classes, attributes
+with defaults can be accessed on the class object itself and any explicit subclasses of the
+protocol class. It cannot be assumed to exist on the meta-type of any arbitrary inhabitant of the
+protocol type, however; an implicit subtype of the protocol will not necessarily have a default value
+for the instance attribute provided in its class body:
+
+```py
+class HasXWithDefault(Protocol):
+    x: int = 42
+
+reveal_type(HasXWithDefault.x)  # revealed: int
+
+class ExplicitSubclass(HasXWithDefault): ...
+
+reveal_type(ExplicitSubclass.x)  # revealed: int
+
+def f(arg: HasXWithDefault):
+    # TODO: should emit `[unresolved-reference]` and reveal `Unknown`
+    reveal_type(type(arg).x)  # revealed: int
+```
+
+Attribute members are allowed to have assignments in methods on the protocol class, just like
+non-protocol classes. Unlike other classes, however, *implicit* instance attributes -- those that
+are not declared in the class body -- are not allowed:
+
+```py
+class Foo(Protocol):
+    x: int
+    y: str
+
+    def __init__(self) -> None:
+        self.x = 42  # fine
+        self.a = 56  # error
+
+    def non_init_method(self) -> None:
+        self.y = 64  # fine
+        self.b = 72  # error
+```
+
+If a protocol has 0 members, then all other types are assignable to it, and all fully static types are subtypes of it:
+
+```py
+from typing import Protocol
+
+class UniversalSet(Protocol): ...
+
+static_assert(is_assignable_to(object, UniversalSet))  # error: [static-assert-error]
+static_assert(is_subtype_of(object, UniversalSet))  # error: [static-assert-error]
```

Comment:
TODO comments? ...I think at this point I'll stop commenting on these, but there are more below...